src/tools.ts

```diff
@@ -28,6 +28,21 @@ function getRangesFromDecorations (
     .map((decoration) => decoration.range)
 }
 
+function getEditableRange (fullModelRange: monaco.Range, ranges: monaco.Range[], withDecoration: boolean): monaco.Range | undefined {
+  if (ranges.length <= 0) {
+    return undefined
+  }
+
+  if (!withDecoration) {
+    return ranges[ranges.length - 1]
+  }
+
+  const firstUneditableRange = ranges[0]
+  return firstUneditableRange != null
+    ? new monaco.Range(fullModelRange.startLineNumber, fullModelRange.startColumn, firstUneditableRange.startLineNumber, firstUneditableRange.startColumn)
+    : undefined
+}
```

Comment:
So we only support a single range now? what is the point of computing multiple one then?

We only have some code to invert the ranges in this file around line 320

Reply:
I added a generic function that recomputes the editor operations: f538ee0.

Reply:
Can the code in the paste callback be removed then?

Reply:
Yes, we should be able to remove the modification of `editor.trigger`. I'll removing it and test it.

Reply:
Here is the commit that cleans up the code: 04bfcfb.

Reply:
I found a bug while testing, here is the fix: c40f3c1.

Reply:
Can you please split the code a little bit more? I'm struggling to understand the logic

Reply:
Here is the refactor:  b598071.

I don't think I can split more the functions.

Reply:
I moved the functions in other files: 8faaeb6.

I changed the logic to split the ranges, then split the operation text separately: c291e16.

Reply:
Here are the commits that fix the last bugs: 8412173 and e47df09.